sebs/aws/triggers.py

```diff
@@ -123,3 +126,171 @@ def serialize(self) -> dict:
+        queue_arn = sqs_client.get_queue_attributes(
+                                                    QueueUrl=queue_url,
+                                                    AttributeNames=["QueueArn"]
+                                                   )["Attributes"]["QueueArn"]
+
+        self.logging.debug("Created queue")
+
+        # Add queue trigger
+        if (not len(lambda_client.list_event_source_mappings(EventSourceArn=queue_arn,
+                                                             FunctionName=self.name)
+                                                        ["EventSourceMappings"])):
+            lambda_client.create_event_source_mapping(
+                EventSourceArn=queue_arn,
+                FunctionName=self.name,
+                MaximumBatchingWindowInSeconds=1
+            )
+
+        # Publish payload to queue
+        sqs_client.send_message(QueueUrl=queue_url, MessageBody=serialized_payload)
+        self.logging.info(f"Sent message to queue {self.name}")
+
+        # TODO(oana): gather metrics
+
+    def async_invoke(self, payload: dict) -> concurrent.futures.Future:
+
+        pool = concurrent.futures.ThreadPoolExecutor()
+        fut = pool.submit(self.sync_invoke, payload)
+        return fut
+
+    def serialize(self) -> dict:
+        return {"type": "Queue", "name": self.name}
+
+    @staticmethod
+    def deserialize(obj: dict) -> Trigger:
+        return QueueTrigger(obj["name"])
+
+
+class StorageTrigger(Trigger):
+    def __init__(self, fname: str, deployment_client: Optional[AWS] = None):
+        super().__init__()
+        self.name = fname
+        self._deployment_client = deployment_client
+
+    @staticmethod
+    def typename() -> str:
+        return "AWS.StorageTrigger"
+
+    @property
+    def deployment_client(self) -> AWS:
+        assert self._deployment_client
+        return self._deployment_client
+
+    @deployment_client.setter
+    def deployment_client(self, deployment_client: AWS):
+        self._deployment_client = deployment_client
+
+    @staticmethod
+    def trigger_type() -> Trigger.TriggerType:
+        return Trigger.TriggerType.STORAGE
+
+    def sync_invoke(self, payload: dict) -> ExecutionResult:
+
+        self.logging.debug(f"Invoke function {self.name}")
+
+        # Init clients
+        lambda_client = self.deployment_client.get_lambda_client()
+        s3 = boto3.resource('s3')
+
+        # Prep
+        serialized_payload = json.dumps(payload)
+        bucket_name = self.name.replace('_', '-')  # AWS disallows underscores in bucket names
+        function_arn = lambda_client.get_function(FunctionName=self.name) \
+                        ["Configuration"]["FunctionArn"]
+
+        # Create bucket
+        self.logging.info(f"Creating bucket {bucket_name}")
+
+        region = self.deployment_client.config.region
+        if (region == "us-east-1"):
+            s3.create_bucket(Bucket=bucket_name)
```

Comment:
I think we should use the existing infrastructure to create buckets (applies to all platforms): https://github.com/spcl/serverless-benchmarks/blob/master/sebs/faas/storage.py#L254

We can either use on of the existing buckets or add a fourth type "trigger". Then, each function will use this one bucket, but the trigger will use a specific path within bucket/container. I think that all platforms support specifying triggers based on the path, although it's less obvious on GCP: https://stackoverflow.com/questions/45404270/is-it-possible-to-scope-firebase-functions-to-a-folder-inside-a-storage-bucket